File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] ans = new int[n];
        int p = 1;
        int zeros = 0;
        int zeroIndex = -1;
        for(int i = 0 ; i < nums.length ; i++){
            if(nums[i] == 0) {
                zeros++;
                zeroIndex = i;
            }
            else p*=nums[i];
        }
        if(zeros > 1) return ans;
        else if(zeros == 1){
            ans[zeroIndex] = p;
        }
        else{
            for(int i = 0 ; i < n ; i++){
                ans[i] = p / nums[i];
            }
        }
        return ans;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        ans = [0] * n
        prefixes = [1] * n
        suffixes = [1] * n
        for i in range(n - 1):
            prefixes[i + 1] *= prefixes[i] * nums[i]
            suffixes[n - 2 - i] *= suffixes[n - 1 - i] * nums[n - 1 - i]
        ans[0], ans[n - 1] = suffixes[0], prefixes[n - 1]
        for i in range(1, n - 1):
            ans[i] = prefixes[i] * suffixes[i]
        return ans
        